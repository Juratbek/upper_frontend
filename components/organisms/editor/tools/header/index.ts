export * from './Header.tool';
export * from './Header.settings';
export * from './Header.types';
