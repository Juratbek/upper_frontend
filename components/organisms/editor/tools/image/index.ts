export * from './Image.tool';
export * from './Image.settings';
export * from './Image.types';
