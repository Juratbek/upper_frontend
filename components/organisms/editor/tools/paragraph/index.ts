export * from './Paragraph.tool';
export * from './Paragraph.types';
