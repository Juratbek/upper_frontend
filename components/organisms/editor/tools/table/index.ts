export * from './Table.tool';
export * from './Table.types';
