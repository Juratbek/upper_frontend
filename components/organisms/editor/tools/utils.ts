import { KeyboardEvent } from 'react';

import { IEditorAPI, IEditorContext } from '../context/EditorContext.types';
import { IBlockData, IBlockNode, TInitialBlockData } from '../instance/Editor.types';
import { isEmpty } from '../utils/html';
import { Selection } from '../utils/selection';

export const generateBlockNode = (element: HTMLDivElement, block: IBlockData): IBlockNode => {
  // just creating new object from params will not work
  // because element is provided using react ref and stores reference to the reference node not to the actual element
  return Object.assign(element, {
    type: block.type,
    id: block.id,
    data: block.data,
  } satisfies IBlockData);
};

export const textBlockKeydownHandler = <T extends { text: string }>(
  event: KeyboardEvent,
  api: IEditorAPI,
  element: HTMLElement,
  blockId: IBlockData['id'],
  config?: { shouldMergeOnBackspace?: boolean },
) => {
  const { code } = event;

  // if user pressed Enter with ctrl or command add new paragraph
  if (code === 'Enter') {
    api.addBlock('paragraph', blockId);
    event.preventDefault();
    return;
  }

  // if paragraph is empty when user presses Back key remove this paragraph
  if (code === 'Backspace' && isEmpty(element)) {
    api
      .removeBlock(blockId)
      .then(({ prevBlocks }) => focusPreviousBlock(prevBlocks, blockId, event));
    return;
  }

  if (code === 'Backspace' && config?.shouldMergeOnBackspace) {
    const selection = Selection.selection;
    if (!selection) return;

    const range = selection.getRangeAt(0);
    const offset = range.startOffset;

    if (offset === 0) {
      api.mergeWithPrevBlock<T>(blockId, (prevBlock, currentBlock) => {
        prevBlock.data.text += ` ${currentBlock.data.text}`;

        return prevBlock.data;
      });
      return;
    }
  }
};

export function getCurrentBlock<T>({ data, hoveredBlock }: IEditorContext<T>) {
  return data.find((b) => hoveredBlock?.id === b.id);
}

function focusPreviousBlock(
  prevBlocks: IBlockData<TInitialBlockData>[],
  blockId: IBlockData['id'],
  event: KeyboardEvent,
) {
  const prevBlock = prevBlocks.find((_, index) => {
    const nextBlock = prevBlocks[index + 1];
    // if next block is the deleted block -> current block is previous block
    if (nextBlock?.id === blockId) return true;

    return false;
  });

  if (!prevBlock) return;

  const prevBlockElement = document.getElementById(prevBlock?.id);
  const textElements = prevBlockElement?.querySelectorAll(
    'h1, h2, h3, h4, h5, h6, p',
  ) as NodeListOf<HTMLParagraphElement | HTMLHeadingElement>;

  if (!textElements || textElements.length < 1) return;

  const lastTextElement = textElements[textElements.length - 1];
  lastTextElement.focus();

  // Create a range (a span of content) and a selection (current user selection)
  const range = document.createRange();
  const selection = window.getSelection();

  if (!selection) return;

  // Move the range to the end of the content inside the element
  range.selectNodeContents(lastTextElement);
  range.collapse(false);

  // Remove any current selections
  selection.removeAllRanges();

  // Add the new range to the selection
  selection.addRange(range);

  event.preventDefault();
}
