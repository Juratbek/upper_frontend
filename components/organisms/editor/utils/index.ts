import { EDITOR_BLOCKS_CONTAINER_ID, EDITOR_ROOT_ID } from '../instance/Editor';

export function getEditorBlocksContainer() {
  return document.getElementById(EDITOR_BLOCKS_CONTAINER_ID);
}

export function getEditorRoot() {
  return document.getElementById(EDITOR_ROOT_ID);
}
