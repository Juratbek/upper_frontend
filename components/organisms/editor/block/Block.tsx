import { memo, useCallback, useRef } from 'react';

import { generateBlockNode } from '../tools/utils';
import cls from './Block.module.scss';
import { IBlockProps } from './Block.types';

export const Block = memo(function Component({ children, onMouseEnter, ...props }: IBlockProps) {
  const blockRef = useRef<HTMLDivElement>(null);
  const mouseEnterHandler = useCallback(() => {
    if (!blockRef.current) return;

    const node = generateBlockNode(blockRef.current, props);

    if (blockRef.current) onMouseEnter?.(node);
  }, [props]);

  const clickHandler = useCallback(
    () => props.onClick?.({ data: props.data, id: props.id, type: props.type }),
    [props],
  );

  const focusHandler = useCallback(
    () => props.onFocus?.({ data: props.data, id: props.id, type: props.type }),
    [props],
  );

  return (
    <div
      className={cls.block}
      ref={blockRef}
      onMouseEnter={mouseEnterHandler}
      onClick={clickHandler}
      onFocus={focusHandler}
    >
      {children}
    </div>
  );
});
