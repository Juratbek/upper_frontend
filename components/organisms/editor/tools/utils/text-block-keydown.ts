import { KeyboardEvent } from 'react';

import { IEditorAPI } from '../../context';
import { IBlockData, TInitialBlockData } from '../../instance/Editor.types';
import { isEmpty } from '../../utils/html';
import { Selection } from '../../utils/selection';
import { IParagraphData } from '../paragraph';

const textNodeNames = 'h1, h2, h3, h4, h5, h6, p';

export const textBlockKeydownHandler = <T extends { text: string }>(
  event: KeyboardEvent,
  api: IEditorAPI,
  element: HTMLElement,
  block: Pick<IBlockData, 'id' | 'type'>,
  config?: { shouldMergeOnBackspace?: boolean },
) => {
  const { code } = event;
  const { id: blockId } = block;

  // if user pressed Enter add new paragraph
  if (code === 'Enter') {
    handleEnter(event, api, element, block);
    return;
  }

  // if paragraph is empty when user presses Back key remove this paragraph
  if (code === 'Backspace' && isEmpty(element)) {
    api
      .removeBlock(blockId)
      .then(({ prevBlocks }) => focusPreviousBlock(prevBlocks, blockId, event));
    return;
  }

  if (code === 'Backspace' && config?.shouldMergeOnBackspace) {
    const range = Selection.range;
    if (!range) return;
    const offset = range.startOffset;

    if (offset === 0) {
      api.mergeWithPrevBlock<T>(blockId, (prevBlock, currentBlock) => {
        prevBlock.data.text += ` ${currentBlock.data.text}`;

        return prevBlock.data;
      });
      return;
    }
  }
};

function handleEnter(
  event: KeyboardEvent,
  api: IEditorAPI,
  element: HTMLElement,
  block: Pick<IBlockData, 'id' | 'type'>,
) {
  const { id: blockId, type: blockType } = block;

  const selection = Selection.selection;
  if (!selection) return;

  const range = selection.getRangeAt(0);
  const cursorPosition = range.startOffset;

  // Clone the range so we can modify it without affecting the current selection
  const cloneRange = range.cloneRange();

  // Create a new range that starts from the cursor position to the end of the paragraph
  cloneRange.setStart(range.startContainer, cursorPosition);
  cloneRange.setEndAfter(element.lastChild!);

  // Extract the contents of the range into a DocumentFragment
  const fragment = cloneRange.extractContents();

  // Convert the fragment to HTML
  const temp = document.createElement('div');
  temp.appendChild(fragment);
  const afterCursorHtml = temp.innerHTML;
  const beforeCursorHtml = element.innerHTML;

  // if there is no text after the cursor -> only add a paragraph
  if (!temp.innerText.trim()) {
    api.addBlock('paragraph', blockId);
    event.preventDefault();
    return;
  }

  // update current block state
  api.setBlock<IParagraphData>({
    id: blockId,
    type: blockType,
    data: { text: beforeCursorHtml },
  });

  // update block content (block will not be rendered when text content is updated)
  const currentBlockElement = document.getElementById(blockId)!;
  const textElement = currentBlockElement.querySelector(textNodeNames)!;
  textElement.innerHTML = beforeCursorHtml;

  // add a new block
  api.addBlock('paragraph', blockId, { text: afterCursorHtml }).then((newBlock) => {
    const newBlockElement = document.getElementById(newBlock.id);
    const paragraph = newBlockElement?.querySelector('p');
    paragraph?.focus();
  });

  event.preventDefault();
}

function getLastTextElement(element: HTMLElement) {
  const textElements = element?.querySelectorAll(textNodeNames) as NodeListOf<
    HTMLParagraphElement | HTMLHeadingElement
  >;

  if (!textElements || textElements.length < 1) return null;

  return textElements[textElements.length - 1];
}

function focusPreviousBlock(
  prevBlocks: IBlockData<TInitialBlockData>[],
  blockId: IBlockData['id'],
  event: KeyboardEvent,
) {
  const prevBlock = prevBlocks.find((_, index) => {
    const nextBlock = prevBlocks[index + 1];
    // if next block is the deleted block -> current block is previous block
    if (nextBlock?.id === blockId) return true;

    return false;
  });

  if (!prevBlock) return;

  const prevBlockElement = document.getElementById(prevBlock?.id);

  const lastTextElement = getLastTextElement(prevBlockElement!);
  if (!lastTextElement) return;
  lastTextElement.focus();

  // Create a range (a span of content) and a selection (current user selection)
  const range = document.createRange();
  const selection = window.getSelection();

  if (!selection) return;

  // Move the range to the end of the content inside the element
  range.selectNodeContents(lastTextElement);
  range.collapse(false);

  // Remove any current selections
  selection.removeAllRanges();

  // Add the new range to the selection
  selection.addRange(range);

  event.preventDefault();
}
