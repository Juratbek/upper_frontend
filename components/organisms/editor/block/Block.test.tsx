import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { vi } from 'vitest';

import { createBlock } from '../context/EditorContext.utils';
import { generateBlockNode } from '../tools/utils';
import { Block } from './Block';
import { IBlockProps } from './Block.types';

const newBlock = createBlock('paragraph');

const mockProps: IBlockProps = {
  id: newBlock.id,
  data: newBlock.data,
  type: 'paragraph',
  children: 'block content',
};

describe('Editor block', () => {
  it('snapshot', () => {
    render(<Block {...mockProps} id='9e02c085' />);
    expect(document.body).toMatchSnapshot();
  });

  it('calls mouse enter callback', async () => {
    const mockMouseEnter = vi.fn();
    mockProps.onMouseEnter = mockMouseEnter;
    render(<Block {...mockProps} />);

    const block = screen.getByTestId('block') as HTMLDivElement;
    userEvent.hover(block);

    const node = generateBlockNode(block, mockProps);

    await waitFor(() => {
      expect(mockMouseEnter).toHaveBeenCalledWith(node);
    });
  });

  it('calls click callback', async () => {
    const mockClick = vi.fn();
    mockProps.onClick = mockClick;
    render(<Block {...mockProps} />);

    const block = screen.getByTestId('block') as HTMLDivElement;
    userEvent.click(block);

    await waitFor(() => {
      expect(mockClick).toHaveBeenCalledWith({
        data: mockProps.data,
        id: mockProps.id,
        type: mockProps.type,
      });
    });
  });

  it('calls focus callback', async () => {
    const mockFocusCb = vi.fn();
    mockProps.onFocus = mockFocusCb;
    render(<Block {...mockProps} />);

    const block = screen.getByTestId('block') as HTMLDivElement;
    fireEvent.focus(block);

    await waitFor(() => {
      expect(mockFocusCb).toHaveBeenCalledWith({
        data: mockProps.data,
        id: mockProps.id,
        type: mockProps.type,
      });
    });
  });
});
