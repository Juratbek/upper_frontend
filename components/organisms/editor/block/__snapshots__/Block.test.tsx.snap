// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Editor block > snapshot 1`] = `
<body>
  <div>
    <div
      class="block"
      data-testid="block"
      id="9e02c085"
    >
      block content
    </div>
  </div>
</body>
`;
